File: ios/FirebaseAuth/Utils/FirebaseAuthHelper.m
/**
 * Copyright 2016 Marcel Piestansky (http://marpies.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FirebaseAuthHelper.h"
#import "FirebaseAuth.h"
#import "FirebaseAuthEvent.h"
#import <AIRExtHelpers/MPStringUtils.h>
#import <FirebaseAuth/FIRUserInfo.h>

@implementation FirebaseAuthHelper

# pragma mark - Public API

- (instancetype) init {
    self = [super init];
    
    [FirebaseAuth log:@"FirebaseAuthHelper::init"];
    
    if( self != nil ) {
        [self addAuthListener];
    }
    
    return self;
}

# pragma mark - Create user / Sign in

- (void) createUserWithEmail:(NSString*) email password:(NSString *)password completion:(FIRAuthResultCallback) completion {
    [FirebaseAuth log:@"Creating new user with email and password"];
    [[FIRAuth auth] createUserWithEmail:email password:password completion:completion];
}

- (void) signInAnonymouslyWithCompletion:(FIRAuthResultCallback) completion {
    [FirebaseAuth log:@"Signing in anonymously"];
    [[FIRAuth auth] signInAnonymouslyWithCompletion:completion];
}

- (void) signInWithEmail:(NSString*) email password:(NSString *)password completion:(FIRAuthResultCallback) completion {
    [FirebaseAuth log:@"Signing in with email and password"];
    [[FIRAuth auth] signInWithEmail:email password:password completion:completion];
}

- (void) signInWithGoogleAccount:(NSString*) idToken accessToken:(NSString*) accessToken completion:(FIRAuthResultCallback) completion {
    [FirebaseAuth log:@"Signing in with Google account"];
    FIRAuthCredential* credential = [FIRGoogleAuthProvider credentialWithIDToken:idToken accessToken:accessToken];
    [[FIRAuth auth] signInWithCredential:credential completion:completion];
}

- (void) signInWithFacebookAccount:(NSString*) accessToken completion:(FIRAuthResultCallback) completion {
    [FirebaseAuth log:@"Signing in with Facebook account"];
    FIRAuthCredential* credential = [FIRFacebookAuthProvider credentialWithAccessToken:accessToken];
    [[FIRAuth auth] signInWithCredential:credential completion:completion];
}

- (void) signInWithGithubAccount:(NSString*) accessToken completion:(FIRAuthResultCallback) completion {
    [FirebaseAuth log:@"Signing in with Github account"];
    FIRAuthCredential* credential = [FIRGitHubAuthProvider credentialWithToken:accessToken];
    [[FIRAuth auth] signInWithCredential:credential completion:completion];
}

- (void) signInWithTwitterAccount:(NSString*) accessToken secret:(NSString*) secret completion:(FIRAuthResultCallback) completion {
    [FirebaseAuth log:@"Signing in with Twitter account"];
    FIRAuthCredential* credential = [FIRTwitterAuthProvider credentialWithToken:accessToken secret:secret];
    [[FIRAuth auth] signInWithCredential:credential completion:completion];
}

# pragma mark - Link / Unlink

- (void) linkWithEmail:(nonnull NSString*) email password:(nonnull NSString *)password completion:(nullable FIRAuthResultCallback)completion {
    [FirebaseAuth log:@"Linking with email and password"];
    FIRAuthCredential* credential = [FIREmailPasswordAuthProvider credentialWithEmail:email password:password];
    [self linkWithCredential:credential completion:completion];
}

- (void) linkWithGoogleAccount:(NSString*) idToken accessToken:(NSString*) accessToken completion:(FIRAuthResultCallback) completion {
    [FirebaseAuth log:@"Linking with Google account"];
    FIRAuthCredential* credential = [FIRGoogleAuthProvider credentialWithIDToken:idToken accessToken:accessToken];
    [self linkWithCredential:credential completion:completion];
}

- (void) linkWithFacebookAccount:(NSString*) accessToken completion:(FIRAuthResultCallback) completion {
    [FirebaseAuth log:@"Linking with Facebook account"];
    FIRAuthCredential* credential = [FIRFacebookAuthProvider credentialWithAccessToken:accessToken];
    [self linkWithCredential:credential completion:completion];
}

- (void) linkWithGithubAccount:(nonnull NSString*) accessToken completion:(nullable FIRAuthResultCallback)completion {
    [FirebaseAuth log:@"Linking with Github account"];
    FIRAuthCredential* credential = [FIRGitHubAuthProvider credentialWithToken:accessToken];
    [self linkWithCredential:credential completion:completion];
}

- (void) linkWithTwitterAccount:(nonnull NSString*) accessToken secret:(nonnull NSString*) secret completion:(nullable FIRAuthResultCallback) completion {
    [FirebaseAuth log:@"Linking with Twitter account"];
    FIRAuthCredential* credential = [FIRTwitterAuthProvider credentialWithToken:accessToken secret:secret];
    [self linkWithCredential:credential completion:completion];
}

- (void) unlinkFromProvider:(NSString*) providerId completion:(FIRAuthResultCallback) completion {
    FIRUser* user = [self getUser];
    if( user != nil ) {
        [FirebaseAuth log:[NSString stringWithFormat:@"Unlinking from provider: %@", providerId]];
        [user unlinkFromProvider:providerId completion:completion];
    } else {
        completion( nil, [NSError errorWithDomain:@"com.marpies.ane.firebase.auth.error"
                                             code:1338
                                         userInfo:@{ NSLocalizedDescriptionKey: @"Unable to unlink from provider, user is not signed in." }] );
    }
}

# pragma mark - User updates

- (void) updateEmail:(NSString*) email completion:(FIRUserProfileChangeCallback) completion {
    FIRUser* user = [self getUser];
    if( user != nil ) {
        [user updateEmail:email completion:completion];
    } else {
        completion( [NSError errorWithDomain:@"com.marpies.ane.firebase.auth.error"
                                        code:1339
                                    userInfo:@{ NSLocalizedDescriptionKey: @"Unable to update email, user is not signed in." }] );
    }
}

# pragma mark - Misc

- (BOOL) signOut {
    NSError *error;
    [[FIRAuth auth] signOut:&error];
    if( error == nil ) {
        return YES;
    }
    return NO;
}

- (FIRUser*) getUser {
    return [FIRAuth auth].currentUser;
}

- (void) processAuthResponse:(nullable FIRUser*) user error:(nullable NSError*) error callbackId:(int) callbackId {
    /* Success */
    if( error == nil ) {
        NSString* userJSON = [self getJSONFromUser:user];
        NSMutableDictionary* response = [NSMutableDictionary dictionary];
        response[@"user"] = userJSON;
        response[@"callbackId"] = @(callbackId);
        NSString* responseJSON = [MPStringUtils getJSONString:response];
        if( responseJSON != nil ) {
            [FirebaseAuth log:@"User authentication success"];
            [FirebaseAuth dispatchEvent:FBA_SIGN_IN_SUCCESS withMessage:responseJSON];
        } else {
            [FirebaseAuth log:@"User authentication success, but failed to create response"];
            [FirebaseAuth dispatchEvent:FBA_SIGN_IN_ERROR withMessage:[MPStringUtils getEventErrorJSONString:callbackId errorMessage:@"Failed to create response."]];
        }
    }
    /* Error */
    else {
        [FirebaseAuth log:[NSString stringWithFormat:@"Error authenticating user: %@", error.localizedDescription]];
        [FirebaseAuth dispatchEvent:FBA_SIGN_IN_ERROR withMessage:[MPStringUtils getEventErrorJSONString:callbackId errorMessage:error.localizedDescription]];
    }
}

- (void) processProfileChangeResponse:(nullable NSError*) error callbackId:(int) callbackId {
    /* Success */
    if( error == nil ) {
        [FirebaseAuth log:@"Successfully changed user profile"];
        [FirebaseAuth dispatchEvent:FBA_PROFILE_CHANGE_SUCCESS withMessage:[NSString stringWithFormat:@"%i", callbackId]];
    }
    /* Error */
    else {
        [FirebaseAuth log:[NSString stringWithFormat:@"Error changing user profile: %@", error.localizedDescription]];
        [FirebaseAuth dispatchEvent:FBA_PROFILE_CHANGE_ERROR withMessage:[MPStringUtils getEventErrorJSONString:callbackId errorMessage:error.localizedDescription]];
    }
}

# pragma mark - Private API

- (NSString*) getJSONFromUser:(FIRUser*) newUser {
    FIRUser* user = (newUser == nil) ? [self getUser] : newUser;
    if( user != nil ) {
        NSMutableDictionary* json = [NSMutableDictionary dictionary];
        json[@"uid"] = user.uid;
        json[@"isAnonymous"] = @(user.isAnonymous);
        if( user.displayName != nil ) {
            json[@"displayName"] = user.displayName;
        }
        if( user.email != nil ) {
            json[@"email"] = user.email;
        }
        if( user.photoURL != nil ) {
            json[@"photoURL"] = user.photoURL;
        }
        NSMutableArray* providerData = [NSMutableArray array];
        for( id<FIRUserInfo> userInfo in user.providerData ) {
            NSString* userInfoJSON = [self getJSONFromUserInfo:userInfo];
            if( userInfoJSON != nil ) {
                [providerData addObject:userInfoJSON];
            }
        }
        if( providerData.count > 0 ) {
            json[@"providerData"] = providerData;
        }
        return [MPStringUtils getJSONString:json];
    }
    return nil;
}

- (void) linkWithCredential:(FIRAuthCredential*) credential completion:(nullable FIRAuthResultCallback)completion {
    FIRUser* user = [self getUser];
    if( user != nil ) {
        [user linkWithCredential:credential completion:completion];
    } else {
        [FirebaseAuth log:@"Linking failed because user is not signed in."];
        if( completion != nil ) {
            completion( nil, [NSError errorWithDomain:@"com.marpies.ane.firebase.auth.error"
                                                 code:1337
                                             userInfo:@{ NSLocalizedDescriptionKey: @"Unable to link account, user is not signed in." }] );
        }
    }
}

- (void) addAuthListener {
    [[FIRAuth auth] addAuthStateDidChangeListener:^(FIRAuth * _Nonnull auth, FIRUser * _Nullable user) {
        if (user != nil) {
            [FirebaseAuth log:@"User has signed in"];
            NSString* userJSON = [self getJSONFromUser:user];
            if( userJSON == nil ) {
                userJSON = @"";
            }
            [FirebaseAuth dispatchEvent:FBA_AUTH_STATE_SIGN_IN withMessage:userJSON];
        } else {
            [FirebaseAuth log:@"No user is currently signed in"];
            /* Dispatch internal event to remove cached FIRUser in AS3 lib */
            [FirebaseAuth dispatchEvent:FBA_AUTH_STATE_SIGN_OFF];
        }
    }];
}

- (NSString*) getJSONFromUserInfo:(id<FIRUserInfo>) userInfo {
    NSMutableDictionary* json = [NSMutableDictionary dictionary];
    NSString* providerID = userInfo.providerID;
    if( providerID != nil ) {
        json[@"providerId"] = providerID;
    }
    NSString* uid = userInfo.uid;
    if( uid != nil ) {
        json[@"uid"] = uid;
    }
    NSString* name = userInfo.displayName;
    if( name != nil ) {
        json[@"displayName"] = name;
    }
    NSString* email = userInfo.email;
    if( email != nil ) {
        json[@"email"] = email;
    }
    NSURL* photoURL = userInfo.photoURL;
    if( photoURL != nil ) {
        json[@"photoURL"] = photoURL.absoluteString;
    }
    return [MPStringUtils getJSONString:json];
}

@end

File: android/src/com/marpies/ane/firebase/auth/utils/FirebaseAuthHelper.java
/*
 * Copyright 2016 Marcel Piestansky (http://marpies.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.marpies.ane.firebase.auth.utils;

import android.net.Uri;
import android.support.annotation.NonNull;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.*;
import com.marpies.ane.firebase.auth.data.FirebaseAuthEvent;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class FirebaseAuthHelper implements FirebaseAuth.AuthStateListener {

	private static FirebaseAuthHelper mInstance;

	public static FirebaseAuthHelper getInstance() {
		if( mInstance == null ) {
			mInstance = new FirebaseAuthHelper();
		}
		return mInstance;
	}

	private FirebaseAuthHelper() { }

	/**
	 *
	 *
	 * Public API
	 *
	 *
	 */

	@Override
	public void onAuthStateChanged( @NonNull FirebaseAuth firebaseAuth ) {
		FirebaseUser user = firebaseAuth.getCurrentUser();
		if( user != null ) {
			AIR.log( "User has signed in" );
			String userJSON = getJSONFromUser( user );
			if( userJSON == null ) {
				userJSON = "{}";
			}
			AIR.dispatchEvent( FirebaseAuthEvent.AUTH_STATE_SIGN_IN, userJSON );
		} else {
			AIR.log( "No user is currently signed in" );
			AIR.dispatchEvent( FirebaseAuthEvent.AUTH_STATE_SIGN_OFF );
		}
	}

	public void signInWithCredential( AuthCredential credential, final int callbackId ) {
		FirebaseAuth.getInstance()
				.signInWithCredential( credential )
				.addOnCompleteListener( new OnCompleteListener<AuthResult>() {
					@Override
					public void onComplete( @NonNull Task<AuthResult> task ) {
						processAuthResponse( task, callbackId );
					}
				} );
	}

	public void linkWithCredential( AuthCredential credential, final int callbackId ) {
		FirebaseUser user = getUser();
		if( user != null ) {
			user.linkWithCredential( credential )
					.addOnCompleteListener( new OnCompleteListener<AuthResult>() {
						@Override
						public void onComplete( @NonNull Task<AuthResult> task ) {
							processAuthResponse( task, callbackId );
						}
					} );
		} else {
			dispatchAuthErrorResponse( "Unable to link with provider, user is not signed in.", callbackId );
		}
	}

	public void reauthenticateWithCredential( AuthCredential credential, final int callbackId ) {
		FirebaseUser user = getUser();
		if( user != null ) {
			user.reauthenticate( credential )
					.addOnCompleteListener( new OnCompleteListener<Void>() {
						@Override
						public void onComplete( @NonNull Task<Void> task ) {
							processProfileChangeResponse( task, callbackId );
						}
					} );
		} else {
			dispatchProfileChangeErrorResponse( "Unable to reauthenticate user, user is not signed in.", callbackId );
		}
	}

	public void processAuthResponse( @NonNull Task<AuthResult> task, int callbackId ) {
		if( task.isSuccessful() ) {
			JSONObject response = new JSONObject();
			try {
				response.put( "user", getJSONFromUser( task.getResult().getUser() ) );
				response.put( "callbackId", callbackId );
				AIR.dispatchEvent( FirebaseAuthEvent.SIGN_IN_SUCCESS, response.toString() );
			} catch( JSONException e ) {
				e.printStackTrace();
				AIR.log( "User authentication success, but failed to create response" );
				dispatchAuthErrorResponse( "Failed to create response.", callbackId );
			}
		} else {
			String errorMessage = (task.getException() != null) ? task.getException().getLocalizedMessage() : "Unknown error.";
			AIR.log( "Error authenticating user: " + errorMessage );
			dispatchAuthErrorResponse( errorMessage, callbackId );
		}
	}

	public void unlinkFromProvider( String providerId, final int callbackId ) {
		FirebaseUser user = getUser();
		if( user != null ) {
			user.unlink( providerId )
					.addOnCompleteListener( new OnCompleteListener<AuthResult>() {
						@Override
						public void onComplete( @NonNull Task<AuthResult> task ) {
							processAuthResponse( task, callbackId );
						}
					} );
		} else {
			dispatchAuthErrorResponse( "Unable to unlink from provider, user is not signed in.", callbackId );
		}
	}

	public void updateEmail( String email, final int callbackId ) {
		FirebaseUser user = getUser();
		if( user != null ) {
			user.updateEmail( email )
					.addOnCompleteListener( new OnCompleteListener<Void>() {
						@Override
						public void onComplete( @NonNull Task<Void> task ) {
							processProfileChangeResponse( task, callbackId );
						}
					} );
		} else {
			dispatchProfileChangeErrorResponse( "Unable to update email, user is not signed in.", callbackId );
		}
	}

	public void updatePassword( String password, final int callbackId ) {
		FirebaseUser user = getUser();
		if( user != null ) {
			user.updatePassword( password )
					.addOnCompleteListener( new OnCompleteListener<Void>() {
						@Override
						public void onComplete( @NonNull Task<Void> task ) {
							processProfileChangeResponse( task, callbackId );
						}
					} );
		} else {
			dispatchProfileChangeErrorResponse( "Unable to update password, user is not signed in.", callbackId );
		}
	}

	public void deleteUser( final int callbackId ) {
		FirebaseUser user = getUser();
		if( user != null ) {
			user.delete()
					.addOnCompleteListener( new OnCompleteListener<Void>() {
						@Override
						public void onComplete( @NonNull Task<Void> task ) {
							processProfileChangeResponse( task, callbackId );
						}
					} );
		} else {
			dispatchProfileChangeErrorResponse( "Unable to delete user, user is not signed in.", callbackId );
		}
	}

	public void updateUserProfile( String displayName, String photoURL, final int callbackId ) {
		FirebaseUser user = getUser();
		if( user != null ) {
			UserProfileChangeRequest.Builder request = new UserProfileChangeRequest.Builder();
			if( displayName != null ) {
				request.setDisplayName( displayName );
			}
			if( photoURL != null ) {
				request.setPhotoUri( Uri.parse( photoURL ) );
			}
			user.updateProfile( request.build() )
					.addOnCompleteListener( new OnCompleteListener<Void>() {
						@Override
						public void onComplete( @NonNull Task<Void> task ) {
							processProfileChangeResponse( task, callbackId );
						}
					} );
		} else {
			dispatchProfileChangeErrorResponse( "Unable to update profile, user is not signed in.", callbackId );
		}
	}

	/**
	 *
	 *
	 * Private API
	 *
	 *
	 */

	private void processProfileChangeResponse( @NonNull Task<Void> task, int callbackId ) {
		if( task.isSuccessful() ) {
			AIR.dispatchEvent( FirebaseAuthEvent.PROFILE_CHANGE_SUCCESS, String.valueOf( callbackId ) );
		} else {
			String errorMessage = (task.getException() != null) ? task.getException().getLocalizedMessage() : "Unknown error.";
			AIR.log( "Error changing user profile: " + errorMessage );
			dispatchProfileChangeErrorResponse( errorMessage, callbackId );
		}
	}

	private void dispatchAuthErrorResponse( String errorMessage, int callbackId ) {
		AIR.dispatchEvent( FirebaseAuthEvent.SIGN_IN_ERROR, StringUtils.getEventErrorJSON( callbackId, errorMessage ) );
	}

	private void dispatchProfileChangeErrorResponse( String errorMessage, int callbackId ) {
		AIR.dispatchEvent( FirebaseAuthEvent.PROFILE_CHANGE_ERROR, StringUtils.getEventErrorJSON( callbackId, errorMessage ) );
	}

	private String getJSONFromUser( FirebaseUser newUser ) {
		FirebaseUser user = (newUser == null) ? getUser() : newUser;
		if( user != null ) {
			JSONObject json = new JSONObject();
			try {
				json.put( "uid", user.getUid() );
				json.put( "isAnonymous", user.isAnonymous() );
				if( user.getDisplayName() != null ) {
					json.put( "displayName", user.getDisplayName() );
				}
				if( user.getEmail() != null ) {
					json.put( "email", user.getEmail() );
				}
				if( user.getPhotoUrl() != null ) {
					json.put( "email", user.getPhotoUrl().toString() );
				}
				JSONArray providerData = new JSONArray();
				for( UserInfo userInfo : user.getProviderData() ) {
					String userInfoJSON = getJSONFromUserInfo( userInfo );
					if( userInfoJSON != null ) {
						providerData.put( userInfoJSON );
					}
				}
				if( providerData.length() > 0 ) {
					json.put( "providerData", providerData );
				}
			} catch( JSONException e ) {
				e.printStackTrace();
			}
			return json.toString();
		}
		return null;
	}

	private String getJSONFromUserInfo( UserInfo userInfo ) {
		JSONObject json = new JSONObject();
		try {
			String providerId = userInfo.getProviderId();
			if( providerId != null ) {
				json.put( "providerId", providerId );
			}
			String uid = userInfo.getUid();
			if( uid != null ) {
				json.put( "uid", uid );
			}
			String displayName = userInfo.getDisplayName();
			if( displayName != null ) {
				json.put( "displayName", displayName );
			}
			String email = userInfo.getEmail();
			if( email != null ) {
				json.put( "email", email );
			}
			Uri photoURL = userInfo.getPhotoUrl();
			if( photoURL != null ) {
				json.put( "photoURL", photoURL.toString() );
			}
			return json.toString();
		} catch( JSONException e ) {
			e.printStackTrace();
		}
		return null;
	}

	private FirebaseUser getUser() {
		return FirebaseAuth.getInstance().getCurrentUser();
	}

}
